File: 2042.cpp
#include <iostream>
#include <vector>
#include <cmath>

#define leftChild(parentNode) parentNode * 2
#define rightChild(parentNode) parentNode * 2 + 1

using namespace std;

vector<long long> segmentTree;
vector<long long> numbers = {-1, };

int numOfNumbers, numOfChanges, numOfCalculations, lengthOfTree;

void input();
void solve();
long long initTreeNode(int node, int from, int to);
long long sum(int node, int from, int to, int left, int right);
void updateTree(int node, int from, int to, int index, long long diff);

int main(){
    input();
    solve();
    return 0;
}

void input(){
    scanf("%d %d %d", &numOfNumbers, &numOfChanges, &numOfCalculations);
    lengthOfTree = 1 << ((int)ceil(log2(numOfNumbers)) + 1);
    segmentTree = vector<long long>(lengthOfTree, 0);
    for(int i = 0; i < numOfNumbers; i++){
        long long number;
        scanf("%lld", &number);
        numbers.push_back(number);
    }
    initTreeNode(1, 1, numOfNumbers);
}

void solve(){
    while(numOfChanges != 0 || numOfCalculations != 0){
        int command;
        scanf("%d", &command);
        if(command == 1 && numOfChanges != 0){
            int index;
            long long value;
            scanf("%d %lld", &index, &value);
            updateTree(1, 1, numOfNumbers, index, value - numbers[index]);
            numOfChanges--;
        }
        else if(command == 2 && numOfCalculations != 0){
            int from, to;
            scanf("%d %d", &from, &to);
            printf("%lld\n", sum(1, 1, numOfNumbers, from, to));
            numOfCalculations--;
        }
    }
}

long long initTreeNode(int node, int from, int to){
    if (from == to)
        segmentTree[node] = numbers[from];
    else{
        int mid = (from + to) / 2;
        segmentTree[node] = initTreeNode(leftChild(node), from, mid) + initTreeNode(rightChild(node), mid + 1, to);
    }
    return segmentTree[node];
}

void updateTree(int node, int from, int to, int index, long long diff){
    if (index < from || index > to)
        return;
    
    segmentTree[node] += diff;
    if (from != to){
        int mid = (from + to) / 2;
        updateTree(leftChild(node), from, mid, index, diff);
        updateTree(rightChild(node), mid+1, to, index, diff);
    }
    else
        numbers[from] += diff;
}

long long sum(int node, int from, int to, int rangeFrom, int rangeTo){
    if (from > rangeTo || to < rangeFrom)
        return 0;
    else if (from >= rangeFrom && to <= rangeTo)
        return segmentTree[node];
    else {
        int mid = (from + to) / 2;
        return sum(leftChild(node), from, mid, rangeFrom, rangeTo) + sum(rightChild(node), mid+1, to, rangeFrom, rangeTo);
    }
}

File: 2042.swift
import Foundation

let input = readLine()!.split(separator: " ").map({Int(String($0))!})
let numOfNumbers = input[0]
let numOfCommands = input[1] + input[2]

var numbers: [Int64] = [0]

for _ in 0..<numOfNumbers {
    numbers.append(Int64(readLine()!)!)
}

var segmentTree: [Int64] = Array(repeating: 0, count: 1 << Int(ceil(log2(Double(numOfNumbers))) + 1) - 1)
let _ = initTree(currentFrom: 1, currentTo: numOfNumbers, currentIndex: 0)

for _ in 0..<numOfCommands {
    let command = readLine()!.split(separator: " ").map({Int64(String($0))!})
    
    if command[0] == 1 {
        let target = Int(command[1])
        let difference = command[2] - numbers[target]
        numbers[target] = command[2]
        update(currentFrom: 1, currentTo: numOfNumbers, target: target, currentIndex: 0, difference: difference)
    } else if command[0] == 2 {
        let from = Int(command[1])
        let to = Int(command[2])
        print(sum(from: from, to: to, currentFrom: 1, currentTo: numOfNumbers, currentIndex: 0))
    }
}

func leftChild(of index: Int) -> Int {
    return index * 2 + 1
}

func rightChild(of index: Int) -> Int {
    return index * 2 + 2
}

func initTree(currentFrom: Int, currentTo: Int, currentIndex: Int) -> Int64 {
    if currentFrom == currentTo {
        segmentTree[currentIndex] = numbers[currentFrom]
    } else {
        let mid = (currentFrom + currentTo) / 2
        segmentTree[currentIndex] = initTree(currentFrom: currentFrom, currentTo: mid, currentIndex: leftChild(of: currentIndex)) + initTree(currentFrom: mid + 1, currentTo: currentTo, currentIndex: rightChild(of: currentIndex))
    }
    return segmentTree[currentIndex]
}

func sum(from: Int, to: Int, currentFrom: Int, currentTo: Int, currentIndex: Int) -> Int64 {
    if from <= currentFrom && currentTo <= to {
        return segmentTree[currentIndex]
    } else if currentTo < from || currentFrom > to {
        return 0
    } else {
        let mid = (currentFrom + currentTo) / 2
        return sum(from: from, to: to, currentFrom: currentFrom, currentTo: mid, currentIndex: leftChild(of: currentIndex)) + sum(from: from, to: to, currentFrom: mid + 1, currentTo: currentTo, currentIndex: rightChild(of: currentIndex))
    }
}

func update(currentFrom: Int, currentTo: Int, target: Int, currentIndex: Int, difference: Int64) {
    if currentFrom <= target && target <= currentTo {
        segmentTree[currentIndex] += difference
        
        if currentFrom != currentTo {
            let mid = (currentFrom + currentTo) / 2
            update(currentFrom: currentFrom, currentTo: mid, target: target, currentIndex: leftChild(of: currentIndex), difference: difference)
            update(currentFrom: mid + 1, currentTo: currentTo, target: target, currentIndex: rightChild(of: currentIndex), difference: difference)
        }
    }
}

